File: spring-app/hello/src/main/java/com/example/hello/entity/PostComment.java
package com.example.hello.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Table(name="app_posts_comments")
public class PostComment {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "post_id", nullable = false)
    private Post post;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_id")
    private PostComment parent;

    @Builder.Default
    @OneToMany(mappedBy = "parent", fetch = FetchType.LAZY)
    @OrderBy("createdAt ASC")
    private List<PostComment> replies = new ArrayList<>();

    @Column(columnDefinition = "TEXT", nullable = false)
    private String content;

    @Column(name="like_count")
    private Integer likeCount;

    @Column(name="created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name="updated_at")
    private LocalDateTime updatedAt;

    @Column(name="delete_at")
    private LocalDateTime deletedAt;

    @PrePersist // 생성 직후 자동으로 아래 함수를 호출시킨다
    protected void onCreate() {
        createdAt = LocalDateTime.now();
    }

    @PreUpdate // 업데이트 직후 자동으로 아래 함수를 호출시킨다
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: flutter-app/lib/models/post_comment.dart
import 'package:commute_mate/models/post.dart';
import 'package:commute_mate/models/user.dart';

class PostComment {
  final int? id;
  final int? userId;
  final int? postId;
  final int? parentId;
  final User? user;
  final Post? post;
  final String content;
  final int likeCount;
  final bool isLiked;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<PostComment> replies;

  PostComment({
    this.id,
    this.userId,
    this.postId,
    this.parentId,
    this.user,
    this.post,
    required this.content,
    this.likeCount = 0,
    this.isLiked = false,
    DateTime? createdAt,
    DateTime? updatedAt,
    this.replies = const [],
  }) : createdAt = createdAt ?? DateTime.now(),
       updatedAt = updatedAt ?? DateTime.now();

  factory PostComment.fromJson(Map<String, dynamic> json) {
    return PostComment(
      id: json['id'] as int?,
      parentId: json['parentId'] as int?,
      user: json['user'] != null ? User.fromJson(json['user']) : null,
      content: json['content'] ?? '',
      likeCount: json['likeCount'] ?? 0,
      isLiked: json['isLiked'] ?? false,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : DateTime.now(),
      replies: json['replies'] != null
          ? (json['replies'] as List)
                .map((e) => PostComment.fromJson(e as Map<String, dynamic>))
                .toList()
          : [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'userId': userId,
      'postId': postId,
      if (parentId != null) 'parentId': parentId,
      if (id != null) 'id': id,
      'content': content,
      'likeCount': likeCount,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  PostComment copyWith({
    int? id,
    int? userId,
    int? postId,
    int? parentId,
    User? user,
    Post? post,
    String? content,
    int? likeCount,
    bool? isLiked,
    DateTime? createdAt,
    DateTime? updatedAt,
    List<PostComment>? replies,
  }) {
    return PostComment(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      postId: postId ?? this.postId,
      parentId: parentId ?? this.parentId,
      user: user ?? this.user,
      post: post ?? this.post,
      content: content ?? this.content,
      likeCount: likeCount ?? this.likeCount,
      isLiked: isLiked ?? this.isLiked,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      replies: replies ?? this.replies,
    );
  }
}
